### PR TITLE
Migrate intermediate catch event processor

### LINKED ISSUE #4472: Migrate Intermediate Catch Event to new BPMN lifecycle processor
**Description**

Refactor the current implementation of the BPMN element to match with the new BPMN lifecycle processors defined in #4333.

### PR DESCRIPTION
**Pull request by korthout**
## Description

Migrate intermediate catch event processor.

## Related issues

closes #4472

### CODE REVIEW

**Comment by korthout** [comment]
Force pushes are merely to resolve a conflict with develop and reformat the code to comply.

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/CatchEventSubscriber.java [hunk 1]
```diff
@@ -35,9 +35,10 @@ public CatchEventSubscriber(final CatchEventBehavior catchEventBehavior) {
       context.raiseIncident(ErrorType.EXTRACT_VALUE_ERROR, e.getMessage());
       return false;
     } catch (final MessageNameException e) {
-      // incident was already raised internally; processing was interrupted at the point the
-      // exception was thrown; no further processing shall take place, that's why we return false
-      // here
+      // processing was interrupted at the point the exception was thrown; no further processing
+      // shall take place, that's why we return false here
+      // todo verify that this is the correct level to raise the incident at
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/CatchEventSubscriber.java:40]: Should we discuss this point? Or, is everything clear? :)

File: engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  public void onTerminating(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.unsubscribeFromEvents(context);
+    stateTransitionBehavior.transitionToTerminated(context);
+  }
+
+  @Override
+  public void onTerminated(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.publishTriggeredBoundaryEvent(context);
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java:92]: Nope. An intermediate catch event may not have a boundary event attached to it :grin:

File: engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    incidentBehavior.resolveIncidents(context);
+    stateTransitionBehavior.onElementTerminated(element, context);
+    stateBehavior.consumeToken(context);
+  }
+
+  @Override
+  public void onEventOccurred(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    eventSubscriptionBehavior.triggerIntermediateEvent(context);
+    stateTransitionBehavior.transitionToCompleting(context);
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java:102]: I think we should take the transition to `completing` only if the event was triggered (i.e. the event trigger was found in `triggerIntermediateEvent`).

File: engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            failure -> incidentBehavior.createIncident(failure, context));
+  }
+
+  @Override
+  public void onActivated(
+      final ExecutableCatchEventElement element, final BpmnElementContext context) {
+    if (element.isNone()) {
+      stateTransitionBehavior.transitionToCompleting(context);
+    }
+  }
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java:59]: An intermediate catch event must have an event definition. I don't see a case when it has no event definition and we complete it immediately. I know it comes from the existing code but I have no idea why :sweat_smile: 
Let's remove it if you agree.

File: engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnEventSubscriptionBehavior.java [hunk 5]
```diff
@@ -238,6 +256,8 @@ private void activateBoundaryEvent(
       return Either.left(
           new Failure(
               e.getMessage(), ErrorType.EXTRACT_VALUE_ERROR, context.getElementInstanceKey()));
+    } catch (final MessageNameException e) {
+      return Either.left(e.getFailure());
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnEventSubscriptionBehavior.java:260]: Good catch :+1:

**Review by saig0 — APPROVED** [review]
Looks good :+1: 

Just a few minor comments.

File: engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java (hunk 4 above)
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java:59]: Ah, I was unaware we didn't allow these. I'd argue that there are use cases for a none intermediate catch event (e.g. milestone events). But allowing those is out of the scope of these changes. Should I create a new issue for this or is there a clear reason why we don't allow none intermediate catch events? I was unable to find GH issues about this topic.

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/CatchEventSubscriber.java (hunk 1 above)
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/CatchEventSubscriber.java:40]: It's not clear for me yet. The current implementation uses the 'normal' elementInstanceKey as scopeKey, but the MessageCorrelationKeyException is raising an incident at the `variablesScopeKey` level. Why is this is different?

File: engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java (hunk 4 above)
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/nwe/event/IntermediateCatchEventProcessor.java:59]: I guess you mean the none **throw** event :stuck_out_tongue_winking_eye: 

I see no reason why we should not implement it. I guess that nobody asked for it yet.

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/CatchEventSubscriber.java (hunk 1 above)
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/CatchEventSubscriber.java:40]: Good question. I guess that we should have the same logic for every expression. Maybe, we forgot something.

**Comment by saig0** [comment]
:+1:

**Pull request merged**